File: sdk/php/src/Model/RestError.php
<?php

namespace KuCoin\UniversalSDK\Model;

use Exception;

/**
 * Class RestError
 * Represents an exception for REST requests.
 */
class RestError extends Exception
{
    /**
     * @var RestResponse|null
     */
    private $response;

    /**
     * @var Exception|null
     */
    private $err;

    public function __construct(?RestResponse $response = null, ?Exception $err = null)
    {
        parent::__construct($err ? $err->getMessage() : 'unknown', 0, $err);
        $this->response = $response;
        $this->err = $err;
    }

    public function __toString()
    {
        if ($this->response) {
            return sprintf(
                'request error, server code: %s, server msg: %s, context msg: %s',
                $this->response->code,
                $this->response->message,
                $this->err ? $this->err->getMessage() : 'unknown'
            );
        }
        return 'request error, ' . ($this->err ? $this->err->__toString() : 'unknown');
    }

    public function getError(): ?Exception
    {
        return $this->err;
    }

    public function getCommonResponse(): ?RestResponse
    {
        return $this->response;
    }
}

File: sdk/java/src/main/java/com/kucoin/universal/sdk/model/RestError.java
package com.kucoin.universal.sdk.model;

import lombok.Getter;

/**
 * Exception representing a REST API error.
 */
@Getter
public class RestError extends Exception {

    private final RestResponse<?> response;
    private final Throwable cause;

    public RestError(RestResponse<?> response, Throwable cause) {
        super(cause != null ? cause.getMessage() : "unknown", cause);
        this.response = response;
        this.cause = cause;
    }

    @Override
    public String toString() {
        if (response != null) {
            return String.format(
                    "request error, server code: %s, server msg: %s, context msg: %s",
                    response.getCode(),
                    response.getMessage(),
                    cause != null ? cause.getMessage() : "unknown"
            );
        }
        return "request error, " + (cause != null ? cause.toString() : "unknown");
    }
}
